### DISCUSSION TITLE
I am unable to construct a database

### DISCUSSION THREAD

**Discussion by RS2009** [simolus3/drift#2586] [Q&A]
```dart
import 'package:drift/drift.dart';
import 'dart:io';
import 'package:drift/drift.dart';
import 'package:drift/native.dart';
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';

part 'database.g.dart';


class Users extends Table {
  TextColumn get name => text()();
  IntColumn get id => integer()();

  @override
  Set<Column> get primaryKey => {id};
}

@DriftDatabase(tables: [Users,Login])
class AppDatabase extends _$AppDatabase {
  AppDatabase(QueryExecutor e) : super(e);

  @override
  int get schemaVersion => 1;

  Future setUsers() async {
    await batch((batch) {
      batch.insertAll(Users, [
        UsersCompanion.insert(name: 'User1'),
        UsersCompanion.insert(name: 'user2')
      ]);
    });
  }

  Future<int> insertuser(UsersCompanion entity) async{
    return await into(Users).insert(entity);
  }
}
```

Hi i am a beginner in app Development ,I want to have a drift database for my application but i am unable to query .
Above is the code i used but setusers in batch.inserAll for Users it is asking me to to add a cast and on adding it is changed into 'Users as TableInfo<Table, User>' and when running it gives " [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: type '_Type' is not a subtype of type 'TableInfo<$UsersTable, dynamic>' in type cast" this error

Can someone please help me with this..

**Comment by simolus3** [comment]
Please put code snippets in triple quotes (\`\`\`) for them to be more readable.

To refer to a table, you use a generated field in the `_$AppDatabase` class. The name of that field is usually the lowercase variant of your table class. So for `Users`, it would be `users`:

```dart
  Future setUsers() async {
    await batch((batch) {
      batch.insertAll(users, [
        UsersCompanion.insert(name: 'User1'),
        UsersCompanion.insert(name: 'user2')
      ]);
    });
  }

  Future<int> insertuser(UsersCompanion entity) async {
    return await into(users).insert(entity);
  }
```

**Reply by RS2009** [reply]
Thanks for answering.
And how to see if my data is inserted in the database or not? should we run any main function in main.dart file?

**Reply by simolus3** [reply]
I'm not sure I understand the question. You can call `setUsers` or `insertuser` wherever you want. When you await the future returned by the method and it completes, the data will be in the database. You can do that in `main.dart` if that should happen whenever your app starts.

**Reply by RS2009** [reply]
Hi ,for suppose if i have 4-5 columns in my table i want to select only 2 of the colums .
How can we do that?

**Reply by simolus3** [reply]
You can use [`selectOnly`](https://drift.simonbinder.eu/api/drift/databaseconnectionuser/selectonly) for that, e.g.

```dart
final query = selectOnly(users)..addColumns([users.id, users.name]);

for (final row in await query.get()) {
  // You can read columns added to the query like this:
  final id = row.read(users.id);
  final name = row.read(users.name);
}
```

**Reply by RS2009** [reply]
How to initialize the data base instance in a sample main.dart file  for example:
I have a simple login page where u need to enter user id ,and it checks against the user id from the database and return true or false?

**Reply by simolus3** [reply]
If you want to, you can just store the drift database in a top-level field and reference that in your app:

```dart
final database = AppDatabase(LazyDatabase(() async {
  // put the database file, called db.sqlite here, into the documents folder
  // for your app.
  final dbFolder = await getApplicationDocumentsDirectory();
  final file = File(p.join(dbFolder.path, 'db.sqlite'));
  return NativeDatabase.createInBackground(file);
}));

void main() {
  // ...
  runApp(...);
}
```

Drift databases are opened when they are first used, so there is no harm in the top-level field.

For Flutter apps, there are other packages like `get_it` or `provider` that provide idioms for widgets to access common state like databases.

**Reply by RS2009** [reply]
Can we load data into database from a csv file?

**Reply by simolus3** [reply]
Drift doesn't have builtin APIs for that, but you can read the CSV with another package, create companions for that and then use drift's insert APIs.

**Reply by RS2009** [reply]
I tried doing that But i couldnt find suitable ones.
Do recommend any?

**Reply by simolus3** [reply]
Well, there's [package:csv](https://pub.dev/packages/csv) for instance. I have used for something different a while back and it seemed to work fine.
